// mongodb-mcp-server/dist/tools/mongodb/create/insertMany.js
import { z } from "zod";
import { DbOperationArgs, MongoDBToolBase } from "../mongodbTool.js";
export class InsertManyTool extends MongoDBToolBase {
    constructor() {
        super(...arguments);
        this.name = "insert-many";
        this.description = "Insert an array of documents into a MongoDB collection";
        this.argsShape = {
            ...DbOperationArgs,
            documents: z
                .array(z.object({}).passthrough().describe("An individual MongoDB document"))
                .describe("The array of documents to insert, matching the syntax of the document argument of db.collection.insertMany()"),
        };
        this.operationType = "create";
    }
    async execute({ database, collection, documents, }) {
        const provider = await this.ensureConnected();
        const result = await provider.insertMany(database, collection, documents);
        return {
            content: [
                {
                    text: `Inserted \`${result.insertedCount}\` document(s) into collection "${collection}"`,
                    type: "text",
                },
                {
                    text: `Inserted IDs: ${Object.values(result.insertedIds).join(", ")}`,
                    type: "text",
                },
            ],
        };
    }
}
