// mongodb-mcp-server/dist/tools/mongodb/read/aggregate.js
import { z } from "zod";
import { DbOperationArgs, MongoDBToolBase } from "../mongodbTool.js";
import { EJSON } from "bson";
import { checkIndexUsage } from "../../../helpers/indexCheck.js";
export const AggregateArgs = {
    pipeline: z.array(z.object({}).passthrough()).describe("An array of aggregation stages to execute"),
};
export class AggregateTool extends MongoDBToolBase {
    constructor() {
        super(...arguments);
        this.name = "aggregate";
        this.description = "Run an aggregation against a MongoDB collection";
        this.argsShape = {
            ...DbOperationArgs,
            ...AggregateArgs,
        };
        this.operationType = "read";
    }
    async execute({ database, collection, pipeline, }) {
        const provider = await this.ensureConnected();
        // Check if aggregate operation uses an index if enabled
        if (this.config.indexCheck) {
            await checkIndexUsage(provider, database, collection, "aggregate", async () => {
                return provider
                    .aggregate(database, collection, pipeline, {}, { writeConcern: undefined })
                    .explain("queryPlanner");
            });
        }
        const documents = await provider.aggregate(database, collection, pipeline).toArray();
        const content = [
            {
                text: `Found ${documents.length} documents in the collection "${collection}":`,
                type: "text",
            },
            ...documents.map((doc) => {
                return {
                    text: EJSON.stringify(doc),
                    type: "text",
                };
            }),
        ];
        return {
            content,
        };
    }
}
