// mongodb-mcp-server/dist/tools/atlas/read/inspectAccessList.js
import { z } from "zod";
import { AtlasToolBase } from "../atlasTool.js";
export class InspectAccessListTool extends AtlasToolBase {
    constructor() {
        super(...arguments);
        this.name = "atlas-inspect-access-list";
        this.description = "Inspect Ip/CIDR ranges with access to your MongoDB Atlas clusters.";
        this.operationType = "read";
        this.argsShape = {
            projectId: z.string().describe("Atlas project ID"),
        };
    }
    async execute({ projectId }) {
        const accessList = await this.session.apiClient.listProjectIpAccessLists({
            params: {
                path: {
                    groupId: projectId,
                },
            },
        });
        if (!accessList?.results?.length) {
            throw new Error("No access list entries found.");
        }
        return {
            content: [
                {
                    type: "text",
                    text: `IP ADDRESS | CIDR | COMMENT
------|------|------
${(accessList.results || [])
                        .map((entry) => {
                        return `${entry.ipAddress} | ${entry.cidrBlock} | ${entry.comment}`;
                    })
                        .join("\n")}`,
                },
            ],
        };
    }
}
