// mongodb-mcp-server/dist/tools/atlas/read/listDBUsers.js
import { z } from "zod";
import { AtlasToolBase } from "../atlasTool.js";
export class ListDBUsersTool extends AtlasToolBase {
    constructor() {
        super(...arguments);
        this.name = "atlas-list-db-users";
        this.description = "List MongoDB Atlas database users";
        this.operationType = "read";
        this.argsShape = {
            projectId: z.string().describe("Atlas project ID to filter DB users"),
        };
    }
    async execute({ projectId }) {
        const data = await this.session.apiClient.listDatabaseUsers({
            params: {
                path: {
                    groupId: projectId,
                },
            },
        });
        if (!data?.results?.length) {
            throw new Error("No database users found.");
        }
        const output = `Username | Roles | Scopes
----------------|----------------|----------------
` +
            data.results
                .map((user) => {
                return `${user.username} | ${formatRoles(user.roles)} | ${formatScopes(user.scopes)}`;
            })
                .join("\n");
        return {
            content: [{ type: "text", text: output }],
        };
    }
}
function formatRoles(roles) {
    if (!roles?.length) {
        return "N/A";
    }
    return roles
        .map((role) => `${role.roleName}${role.databaseName ? `@${role.databaseName}${role.collectionName ? `:${role.collectionName}` : ""}` : ""}`)
        .join(", ");
}
function formatScopes(scopes) {
    if (!scopes?.length) {
        return "All";
    }
    return scopes.map((scope) => `${scope.type}:${scope.name}`).join(", ");
}
