// mongodb-mcp-server/dist/common/atlas/cluster.js
import logger, { LogId } from "../logger.js";
export function formatFlexCluster(cluster) {
    return {
        name: cluster.name,
        instanceType: "FLEX",
        instanceSize: undefined,
        state: cluster.stateName,
        mongoDBVersion: cluster.mongoDBVersion,
        connectionString: cluster.connectionStrings?.standardSrv || cluster.connectionStrings?.standard,
    };
}
export function formatCluster(cluster) {
    const regionConfigs = (cluster.replicationSpecs || [])
        .map((replicationSpec) => (replicationSpec.regionConfigs || []))
        .flat()
        .map((regionConfig) => {
        return {
            providerName: regionConfig.providerName,
            instanceSize: regionConfig.electableSpecs?.instanceSize ||
                regionConfig.readOnlySpecs?.instanceSize ||
                regionConfig.analyticsSpecs?.instanceSize,
        };
    });
    const instanceSize = regionConfigs[0]?.instanceSize ?? "UNKNOWN";
    const clusterInstanceType = instanceSize == "M0" ? "FREE" : "DEDICATED";
    return {
        name: cluster.name,
        instanceType: clusterInstanceType,
        instanceSize: clusterInstanceType == "DEDICATED" ? instanceSize : undefined,
        state: cluster.stateName,
        mongoDBVersion: cluster.mongoDBVersion,
        connectionString: cluster.connectionStrings?.standardSrv || cluster.connectionStrings?.standard,
    };
}
export async function inspectCluster(apiClient, projectId, clusterName) {
    try {
        const cluster = await apiClient.getCluster({
            params: {
                path: {
                    groupId: projectId,
                    clusterName,
                },
            },
        });
        return formatCluster(cluster);
    }
    catch (error) {
        try {
            const cluster = await apiClient.getFlexCluster({
                params: {
                    path: {
                        groupId: projectId,
                        name: clusterName,
                    },
                },
            });
            return formatFlexCluster(cluster);
        }
        catch (flexError) {
            const err = flexError instanceof Error ? flexError : new Error(String(flexError));
            logger.error(LogId.atlasInspectFailure, "inspect-cluster", `error inspecting cluster: ${err.message}`);
            throw error;
        }
    }
}
