// mongodb-mcp-server/dist/tools/mongodb/metadata/collectionSchema.js
import { DbOperationArgs, MongoDBToolBase } from "../mongodbTool.js";
import { getSimplifiedSchema } from "mongodb-schema";
export class CollectionSchemaTool extends MongoDBToolBase {
    constructor() {
        super(...arguments);
        this.name = "collection-schema";
        this.description = "Describe the schema for a collection";
        this.argsShape = DbOperationArgs;
        this.operationType = "metadata";
    }
    async execute({ database, collection }) {
        const provider = await this.ensureConnected();
        const documents = await provider.find(database, collection, {}, { limit: 5 }).toArray();
        const schema = await getSimplifiedSchema(documents);
        const fieldsCount = Object.entries(schema).length;
        if (fieldsCount === 0) {
            return {
                content: [
                    {
                        text: `Could not deduce the schema for "${database}.${collection}". This may be because it doesn't exist or is empty.`,
                        type: "text",
                    },
                ],
            };
        }
        return {
            content: [
                {
                    text: `Found ${fieldsCount} fields in the schema for "${database}.${collection}"`,
                    type: "text",
                },
                {
                    text: JSON.stringify(schema),
                    type: "text",
                },
            ],
        };
    }
}
