// mongodb-mcp-server/dist/tools/mongodb/metadata/dbStats.js
import { DbOperationArgs, MongoDBToolBase } from "../mongodbTool.js";
import { EJSON } from "bson";
export class DbStatsTool extends MongoDBToolBase {
    constructor() {
        super(...arguments);
        this.name = "db-stats";
        this.description = "Returns statistics that reflect the use state of a single database";
        this.argsShape = {
            database: DbOperationArgs.database,
        };
        this.operationType = "metadata";
    }
    async execute({ database }) {
        const provider = await this.ensureConnected();
        const result = await provider.runCommandWithCheck(database, {
            dbStats: 1,
            scale: 1,
        });
        return {
            content: [
                {
                    text: `Statistics for database ${database}`,
                    type: "text",
                },
                {
                    text: EJSON.stringify(result),
                    type: "text",
                },
            ],
        };
    }
}
