// mongodb-mcp-server/dist/tools/mongodb/metadata/listCollections.js
import { DbOperationArgs, MongoDBToolBase } from "../mongodbTool.js";
export class ListCollectionsTool extends MongoDBToolBase {
    constructor() {
        super(...arguments);
        this.name = "list-collections";
        this.description = "List all collections for a given database";
        this.argsShape = {
            database: DbOperationArgs.database,
        };
        this.operationType = "metadata";
    }
    async execute({ database }) {
        const provider = await this.ensureConnected();
        const collections = await provider.listCollections(database);
        if (collections.length === 0) {
            return {
                content: [
                    {
                        type: "text",
                        text: `No collections found for database "${database}". To create a collection, use the "create-collection" tool.`,
                    },
                ],
            };
        }
        return {
            content: collections.map((collection) => {
                return {
                    text: `Name: "${collection.name}"`,
                    type: "text",
                };
            }),
        };
    }
}
