// @pulumi/mongodbatlas/getStreamAccountDetails.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStreamAccountDetailsOutput = exports.getStreamAccountDetails = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("./utilities");
/**
 * ## # Data Source: mongodbatlas.getStreamAccountDetails
 *
 * `mongodbatlas.getStreamAccountDetails` returns the AWS Account ID/Azure Subscription ID, and the AWS VPC ID/Azure Virtual Network Name for the group, cloud provider, and region that you specify.
 *
 * ## Example Usage
 *
 * ### S
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as mongodbatlas from "@pulumi/mongodbatlas";
 *
 * const accountDetails = mongodbatlas.getStreamAccountDetails({
 *     projectId: projectId,
 *     cloudProvider: "aws",
 *     regionName: "US_EAST_1",
 * });
 * export const awsAccountId = accountDetails.then(accountDetails => accountDetails.awsAccountId);
 * export const cidrBlock = accountDetails.then(accountDetails => accountDetails.cidrBlock);
 * export const cloudProvider = accountDetails.then(accountDetails => accountDetails.cloudProvider);
 * export const vpcId = accountDetails.then(accountDetails => accountDetails.vpcId);
 * ```
 */
function getStreamAccountDetails(args, opts) {
    opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts || {});
    return pulumi.runtime.invoke("mongodbatlas:index/getStreamAccountDetails:getStreamAccountDetails", {
        "cloudProvider": args.cloudProvider,
        "projectId": args.projectId,
        "regionName": args.regionName,
    }, opts);
}
exports.getStreamAccountDetails = getStreamAccountDetails;
/**
 * ## # Data Source: mongodbatlas.getStreamAccountDetails
 *
 * `mongodbatlas.getStreamAccountDetails` returns the AWS Account ID/Azure Subscription ID, and the AWS VPC ID/Azure Virtual Network Name for the group, cloud provider, and region that you specify.
 *
 * ## Example Usage
 *
 * ### S
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as mongodbatlas from "@pulumi/mongodbatlas";
 *
 * const accountDetails = mongodbatlas.getStreamAccountDetails({
 *     projectId: projectId,
 *     cloudProvider: "aws",
 *     regionName: "US_EAST_1",
 * });
 * export const awsAccountId = accountDetails.then(accountDetails => accountDetails.awsAccountId);
 * export const cidrBlock = accountDetails.then(accountDetails => accountDetails.cidrBlock);
 * export const cloudProvider = accountDetails.then(accountDetails => accountDetails.cloudProvider);
 * export const vpcId = accountDetails.then(accountDetails => accountDetails.vpcId);
 * ```
 */
function getStreamAccountDetailsOutput(args, opts) {
    opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts || {});
    return pulumi.runtime.invokeOutput("mongodbatlas:index/getStreamAccountDetails:getStreamAccountDetails", {
        "cloudProvider": args.cloudProvider,
        "projectId": args.projectId,
        "regionName": args.regionName,
    }, opts);
}
exports.getStreamAccountDetailsOutput = getStreamAccountDetailsOutput;
