// mongodb-mcp-server/dist/tools/mongodb/connect/connect.js
import { z } from "zod";
import { MongoDBToolBase } from "../mongodbTool.js";
import assert from "assert";
const disconnectedSchema = z
    .object({
    connectionString: z.string().describe("MongoDB connection string (in the mongodb:// or mongodb+srv:// format)"),
})
    .describe("Options for connecting to MongoDB.");
const connectedSchema = z
    .object({
    connectionString: z
        .string()
        .optional()
        .describe("MongoDB connection string to switch to (in the mongodb:// or mongodb+srv:// format)"),
})
    .describe("Options for switching the current MongoDB connection. If a connection string is not provided, the connection string from the config will be used.");
const connectedName = "switch-connection";
const disconnectedName = "connect";
const connectedDescription = "Switch to a different MongoDB connection. If the user has configured a connection string or has previously called the connect tool, a connection is already established and there's no need to call this tool unless the user has explicitly requested to switch to a new instance.";
const disconnectedDescription = "Connect to a MongoDB instance";
export class ConnectTool extends MongoDBToolBase {
    constructor(session, config, telemetry) {
        super(session, config, telemetry);
        this.name = disconnectedName;
        this.description = disconnectedDescription;
        // Here the default is empty just to trigger registration, but we're going to override it with the correct
        // schema in the register method.
        this.argsShape = {
            connectionString: z.string().optional(),
        };
        this.operationType = "connect";
        session.on("disconnect", () => {
            this.updateMetadata();
        });
    }
    async execute({ connectionString }) {
        switch (this.name) {
            case disconnectedName:
                assert(connectionString, "Connection string is required");
                break;
            case connectedName:
                connectionString ?? (connectionString = this.config.connectionString);
                assert(connectionString, "Cannot switch to a new connection because no connection string was provided and no default connection string is configured.");
                break;
        }
        await this.connectToMongoDB(connectionString);
        this.updateMetadata();
        return {
            content: [{ type: "text", text: "Successfully connected to MongoDB." }],
        };
    }
    register(server) {
        if (super.register(server)) {
            this.updateMetadata();
            return true;
        }
        return false;
    }
    updateMetadata() {
        if (this.config.connectionString || this.session.serviceProvider) {
            this.update?.({
                name: connectedName,
                description: connectedDescription,
                inputSchema: connectedSchema,
            });
        }
        else {
            this.update?.({
                name: disconnectedName,
                description: disconnectedDescription,
                inputSchema: disconnectedSchema,
            });
        }
    }
}
